### PR TITLE
Decision: User and Database Management

### PR DESCRIPTION
**Pull request by ccremer**
## Summary

* While thinking about the API specs and their implementation, this topic came up which clearly has an impact how the user orders service instances.

Also relevant: #14 

## Checklist

- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog
- [x] Link this PR to related issues.

### CODE REVIEW

**Comment by ccremer** [comment]
Note for review: PR currently is based on #14 , I'll rebase once the other is merged.

Requesting review also from @tobru , since User management affects the UX. Please provide the requirements from the Product's perspective.

**Comment by Kidswiss** [comment]
I personally like proposal 3, but for a first iteration proposal 1 is sufficient.

**Comment by tobru** [comment]
Thanks for the proposals.

For the time being and for the first iteration I'd go with "Proposal 2: Exposed Superuser". With "Proposal 1: Single Non-Privileged User and Database" we will not be able to meet many requirements. Let's use Odoo as an example: Odoo wants to manage the database and if it can't, it's a mess. I don't fear the documented risk of users doing harmful things, if they do, we're here to help (not for free).

For a later iteration, I'd go for "Proposal 4: Manage Users and Databases in another CR", this feels more in line with GitOps and the Kubernetes resource model. "Proposal 3: Manage Users and Databases in the same CR" packages all in one CR. If you want to create a DB user for your preview deployments, you would have to fiddle in the main resource. With Proposal 4 you can just bring your applications requirements with the deployment and also remove it again after the preview deployment gets removed.

**Comment by ZuGao** [comment]
Proposal 4 seems the most sophisticated and we already have experience with such implementation in stardog user role operator which worked out well. The only problem is the engineering that is required for a robust integration between the operator and the postgresql user management. For instance we have to make sure when a user/role is being modified manually on a SQL client, the operator must revert those changes immediately.  For starters I would agree with Tobru that we should start small with proposal 1 or 2 and build from there.

**Comment by ccremer** [comment]
> Let's use Odoo as an example: Odoo wants to manage the database and if it can't, it's a mess.

> The only problem is the engineering that is required for a robust integration between the operator and the postgresql user management. For instance we have to make sure when a user/role is being modified manually on a SQL client, the operator must revert those changes immediately.

I see a conflict here. On one hand I agree we should fulfill the "reconciliation contract" (aka cleanup unknown resources), OTOH removing unknown databases makes (probably legitimate) uses cases with apps that bring their own db/user management impossible.

At some point I could imagine a `userManagement` field in the service instance spec, which could be either `Manual` (expose superuser and don't do anything else) or `Declarative` (use CRs to manage users and db). Just an idea of the direction.

Also, please bear in mind that going for proposal 1 and then to another will bring some migration challenges when releasing a new version of the service/operator.

**Comment by ccremer** [comment]
I've rephrased the page and documented proposal 2 with rationale.
Please have a look again and approve/give feedback

**Comment by Kidswiss** [comment]
I just had a thought:

Maybe we could do it like Crunchydata handles it. By default, there are only unprivileged users. But you can explicitly enable the superuser if you wish.

So I have a new proposal:
By default just act like proposal 1. If the user enables a `superuser` flag or something like that on the instance, then act like proposal 2.

**Comment by ccremer** [comment]
> By default just act like proposal 1. If the user enables a `superuser` flag or something like that on the instance, then act like proposal 2.

What I don't quite like in this idea is that we introduce a field that may be potentially become obsolete when going for Proposal 3 or 4, and that will introduce a breaking API change (or at least an unused, deprecated field).

I think for now the simplest solution is to expose both a non-privileged user and db with same name as the instance, as well as the superuser. So the user can choose which credential to use in their app.

**Comment by Kidswiss** [comment]
> What I don't quite like in this idea is that we introduce a field that may be potentially become obsolete when going for Proposal 3 or 4, and that will introduce a breaking API change (or at least an unused, deprecated field).

Not necessarily, with the Crunchydata operator you can define arbitrary users and databases. But you still have to enable the superuser `postgres` explicitly.

File: docs/modules/ROOT/pages/explanations/decision-usermanagement.adoc
Comment (ZuGao) [inline docs/modules/ROOT/pages/explanations/decision-usermanagement.adoc:144]: Some typos.

**Comment by ZuGao** [comment]
I would also lean towards having Solution 1 and 2 like @Kidswiss proposed. I don't think the superuser field will get obsolete with solution 3 and 4. We may actually leave superuser out of the CRs from Solution 3 and 4 and manage it as we would do in Solution 2.

**Comment by ccremer** [comment]
Ok, I guess it's safe enough to introduce such a field. In any case, _how_ this superuser credential is exposed doesn't really matter in this PR, the relevant part is _that_ we want to expose it.
The exact API fields discussion should be in another PR.

If you're okay with that, I'll gladly await your PR approvals :)

**Review by tobru — APPROVED** [review]

**Pull request merged**
